document_ai_warehouse/common/src/common/utils/document_ai_utils.py

```diff
@@ -0,0 +1,312 @@
+
+            Logger.info(
+                f"batch_extraction - Handling DocAI results for {input_gcs_source} using "
+                f"process output {output_gcs_destination}"
+            )
+            # Get List of Document Objects from the Output Bucket
+            output_blobs = storage_client.list_blobs(
+                output_bucket, prefix=output_prefix + "/"
+            )
+
+            # Document AI may output multiple JSON files per source file
+            # Sharding happens when the output JSON File gets over a size threshold,
+            # like 10MB or something. I have seen it happen around 30 pages, but more
+            # often around 40 or 50 pages.
+            for blob in output_blobs:
+                # Document AI should only output JSON files to GCS
+                if ".json" not in blob.name:
+                    Logger.warning(
+                        f"batch_extraction - Skipping non-supported file: {blob.name} - Mimetype: {blob.content_type}"
+                    )
+                    continue
+
+                if input_gcs_source not in documents.keys():
+                    documents[input_gcs_source] = []
+                documents[input_gcs_source].append(f"gs://{output_bucket}/{blob.name}")
+
+        result = {}
+        for doc in documents:
+            result[doc] = merge_json_files(documents[doc])
+
+        return result
+
+
+def merge_json_files(files):
+    """Merges n json files in memory.
+
+    Args:
+      files: A list of paths to json files.
+
+    Returns:
+      A dictionary containing the merged data.
+    """
+
+    result = {}
+    json_list = []
+
+    for file in files:
+        bucket_name, prefix = split_uri_2_bucket_prefix(file)
+        bucket = storage_client.get_bucket(bucket_name)
+        blob = bucket.blob(prefix)
+        json_str = blob.download_as_string()
+        json_list.append(json.loads(json_str.decode("utf8")))
+
+    for d in json_list:
+        result.update(d)
+
+    doc_json = json.dumps(result)
+
+    document = documentai.Document.from_json(doc_json, ignore_unknown_fields=True)
+    return document
+
+
+# Handling Nested labels for CDE processor
+def get_key_values_dic(
+    entity: documentai.Document.Entity,
+    document_entities: Dict[str, Any],
+    parent_key: str = None,
+) -> None:
+    # Fields detected. For a full list of fields for each processor see
+    # the processor documentation:
+    # https://cloud.google.com/document-ai/docs/processors-list
+
+    entity_key = entity.get("type", "").replace("/", "_")
+    confidence = entity.get("confidence")
+    normalized_value = entity.get("normalizedValue")
+
+    if normalized_value:
+        if (
+            isinstance(normalized_value, dict)
+            and "booleanValue" in normalized_value.keys()
+        ):
+            normalized_value = normalized_value.get("booleanValue")
+        else:
+            normalized_value = normalized_value.get("text")
```

Comment:
I think this is the only part that's not included in the Toolbox currently. This could be a contribution to make if it's needed